pkg/antctl/raw/migrate/convertnetworkpolicy/command.go

```diff
@@ -0,0 +1,488 @@
+}
+
+func convertCalicoServiceToAntreaService(rule v3.Rule) []v1beta1.PeerService {
+	if rule.Destination.Services != nil {
+		return []v1beta1.PeerService{
+			{
+				Name:      rule.Destination.Services.Name,
+				Namespace: rule.Destination.Services.Namespace,
+			},
+		}
+	}
+	return nil
+}
+
+func convertEntityRuleToAntreaNetworkPolicyPeer(entityRule v3.EntityRule) ([]v1beta1.NetworkPolicyPeer, error) {
+	var antreaNetworkPolicyPeers []v1beta1.NetworkPolicyPeer
+	antreaNetworkPolicyPeer := v1beta1.NetworkPolicyPeer{}
+	var err error
+	if antreaNetworkPolicyPeer.PodSelector, err = convertCalicoSelectorToLabelSelector(entityRule.Selector); err != nil {
+		return nil, err
+	}
+	if antreaNetworkPolicyPeer.NamespaceSelector, err = convertCalicoSelectorToLabelSelector(entityRule.NamespaceSelector); err != nil {
+		return nil, err
+	}
+	antreaNetworkPolicyPeers = append(antreaNetworkPolicyPeers, antreaNetworkPolicyPeer)
+	for _, cidr := range entityRule.Nets {
+		antreaNetworkPolicyPeers = append(antreaNetworkPolicyPeers, v1beta1.NetworkPolicyPeer{
+			IPBlock: &v1beta1.IPBlock{
+				CIDR: cidr,
+			},
+		})
+	}
+	return antreaNetworkPolicyPeers, nil
+}
+
+func convertSelectorToAntreaAppliedTo(selector string, namespaceSelector string) ([]v1beta1.AppliedTo, error) {
+	appliedTo := &v1beta1.AppliedTo{}
+	var err error
+	if appliedTo.PodSelector, err = convertCalicoSelectorToLabelSelector(selector); err != nil {
+		return nil, err
+	}
+	if appliedTo.NamespaceSelector, err = convertCalicoSelectorToLabelSelector(namespaceSelector); err != nil {
+		return nil, err
+	}
+
+	return []v1beta1.AppliedTo{*appliedTo}, nil
+}
+
+// convertCalicoSelectorToLabelSelector parses the Calico selector string to a LabelSelector in following format:
+// 1. "k == v" => LabelSelector {MatchLabels: map[string]string{"k": "v"}}
+// 2. "k == v && k2 == v2" => LabelSelector {MatchLabels: map[string]string{"k": "v", "k2": "v2"}}
+// 3. has(abc) => LabelSelector {MatchExpressions: []LabelSelectorRequirement{{Key: "abc", Operator: Exists}}}
+// 4. !has(abc) => LabelSelector {MatchExpressions: []LabelSelectorRequirement{{Key: "abc", Operator: DoesNotExist}}}
+// 5. "k in {v1, v2}" => LabelSelector {MatchExpressions: []LabelSelectorRequirement{{Key: "k", Operator: In, Values: []string{"v1", "v2"}}}}
+// 6. "k not in {v1, v2}" => LabelSelector {MatchExpressions: []LabelSelectorRequirement{{Key: "k", Operator: NotIn, Values: []string{"v1", "v2"}}}}
+// 7. "a != b" => LabelSelector {MatchExpressions: []LabelSelectorRequirement{{Key: "a", Operator: NotIn, Values: []string{"b"}}}}
+// 8. "a == b && c != d" => LabelSelector {MatchLabels: map[string]string{"a": "b"}, MatchExpressions: []LabelSelectorRequirement{{Key: "c", Operator: NotIn, Values: []string{"d"}}}}
+func convertCalicoSelectorToLabelSelector(selector string) (*metav1.LabelSelector, error) {
+	selector = strings.ReplaceAll(
+		strings.ReplaceAll(selector, " ", ""), "'", "")
+	conditions := strings.Split(selector, "&&")
```

Comment:
There should be some selectors are using "||" to concentrate? eg: `<expression 1> || <expression 2>`

Reply:
In Calico NP, there are three kinds of selectors, `spec.selector`, `spec.namespaceSelector` and `spec.serviceAccountSelector`. The intersection of these selectors is the target Pods. So the Calico selector should be indicated as

```bash
CalicoSelector=(selector) && (namespaceSelector) && (serviceAccountSelector)
```

If we allow or expression in selectors, the Calico selector could be indicated as

```bash
CalicoSelector=(aaa || bbb ) && ( ccc || ddd ) && (eee || fff )
```

Then we need an extra expression parser as the length of the selector could be very long, which will result in a very complicated expression.